avr: use new cross compilation infrastructure

matthewbauer opened the pull request:
Gets rid of:
-  avrbinutils
-  avrgcc

to replace with:
-  pkgsCross.avr.buildPackages.binutils
-  pkgsCross.avr.buildPackages.gcc

matthewbauer committed:
avr: use new compilation infrastructure

Gets rid of:
  avrbinutils
  avrgcc

to replace with:
  pkgsCross.avr.buildPackages.binutils
  pkgsCross.avr.buildPackages.gcc

matthewbauer committed:
treewide: use gnu toolchain for avr/arm-embedded

These packages should in theory work with our GCC toolchains, but
there are some definite breakages that need to be tracked down.
Comparing output of these to old gcc-arm-embedded is important.

Affected packages include:

- axolooti
- avrdudess
- opentx
- microscheme
- betaflight
- inav
- blackmagic
- simavr
- gnuk

dtzWill commented:
:+1: :heart:

Ericson2314 reviewed (approved):
Gorgeous!

matthewbauer committed:
systems: add avrlibc for avr systems

matthewbauer committed:
gcc: support avr

- respect libc’s incdir and libdir
- make non-unix systems single threaded
- set LIMITS_H_TEST to false for avr
- misc updates to support new libc’s
- use multilib with avr

For threads we want to use:
- posix on unix systems
- win32 on windows
- single on everything else

For avr:
- add library directories for avrlibc
- to disable relro and bind
- avr5 should have precedence over avr3 - otherwise gcc uses the wrong one

GrahamcOfBorg commented:
<!--REQUEST_ID=47223cbd-66f2-41cb-8a3d-a61ddb7833d3-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=abb83093-6f16-4e58-8ecb-e455951918ac)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on aarch64-linux: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/klgxdd526njckk7cq8j1pxakb2p6wv70-gcc-wrapper-7.3.0
/nix/store/m6llsl31vkhkhxfy055m44444nz9s9ki-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=47223cbd-66f2-41cb-8a3d-a61ddb7833d3-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=5265eeda-42b5-472e-ba39-9d1a8221e07e)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on x86_64-linux: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/10yq7kwlvbc6h658izmrlsspry1g9f3c-gcc-wrapper-7.3.0
/nix/store/glnk048cqva0ig8yrjfw140z2r2v18b4-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=47223cbd-66f2-41cb-8a3d-a61ddb7833d3-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=620ecf79-c25a-4717-8fad-2f12175c7f48)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on x86_64-darwin: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  /nix/store/dnyr1mi70mxsyw865b7307544m8szrh0-mpfr-4.0.1
  /nix/store/f31cpaviz7r983b9sdkkm8crmhsmhad6-expand-response-params
  /nix/store/kjk5rnhcnh6r65fmcfyll5wlbfggb3bn-libmpc-1.1.0
copying path '/nix/store/f31cpaviz7r983b9sdkkm8crmhsmhad6-expand-response-params' from 'https://cache.nixos.org'...
copying path '/nix/store/dnyr1mi70mxsyw865b7307544m8szrh0-mpfr-4.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/kjk5rnhcnh6r65fmcfyll5wlbfggb3bn-libmpc-1.1.0' from 'https://cache.nixos.org'...
copying path '/nix/store/6d95j619zw7n6sbk5dqs36gwrsvif9x0-gcc-7.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/10n6qiaz1piz0gsh05x5i062cscrwi6f-gcc-wrapper-7.3.0' from 'https://cache.nixos.org'...
/nix/store/10n6qiaz1piz0gsh05x5i062cscrwi6f-gcc-wrapper-7.3.0
/nix/store/7l83571b34mmbr1zkxa596vw2wdqxxnh-stdenv-darwin
```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build pkgsCross.avr.hello

GrahamcOfBorg commented:
<!--REQUEST_ID=dc0c3ebc-be37-4b03-877d-8be702b0b085-->
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=3e04609b-547e-4a46-b13a-04fc33e73d6e)

The following builds were skipped because they don't evaluate on x86_64-darwin: pkgsCross.avr.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgsCross.avr.hello' because:
error: attribute 'avr' in selection path 'pkgsCross.avr.hello' not found

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=dc0c3ebc-be37-4b03-877d-8be702b0b085-->
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=754eeaa8-57f8-4485-ab04-63de30db9ddf)

The following builds were skipped because they don't evaluate on x86_64-linux: pkgsCross.avr.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgsCross.avr.hello' because:
error: attribute 'avr' in selection path 'pkgsCross.avr.hello' not found

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=dc0c3ebc-be37-4b03-877d-8be702b0b085-->
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=5911775a-d0fb-4861-a9b9-3b844eb8b396)

The following builds were skipped because they don't evaluate on aarch64-linux: pkgsCross.avr.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgsCross.avr.hello' because:
error: attribute 'avr' in selection path 'pkgsCross.avr.hello' not found

```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build pkgsCross.arduino-avr.hello

I need to figure out a way that we don't break qmk_firmware, which looks like it uses Nix:

https://github.com/qmk/qmk_firmware

GrahamcOfBorg commented:
<!--REQUEST_ID=0048a2ab-6808-47b4-8d28-b32b8fc4728c-->
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=b5c8284f-a5ae-4c95-a6e0-7d0541ac4ae7)

The following builds were skipped because they don't evaluate on x86_64-darwin: pkgsCross.arduino-avr.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgsCross.arduino-avr.hello' because:
error: attribute 'arduino-avr' in selection path 'pkgsCross.arduino-avr.hello' not found

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=0048a2ab-6808-47b4-8d28-b32b8fc4728c-->
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=bb2e6bde-f620-4f27-9d4a-3a8f1057cc85)

The following builds were skipped because they don't evaluate on aarch64-linux: pkgsCross.arduino-avr.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgsCross.arduino-avr.hello' because:
error: attribute 'arduino-avr' in selection path 'pkgsCross.arduino-avr.hello' not found

```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build pkgsCross.arduino-uno.hello

GrahamcOfBorg commented:
<!--REQUEST_ID=0048a2ab-6808-47b4-8d28-b32b8fc4728c-->
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=5d397566-101a-404b-a7d9-4f0689859b56)

The following builds were skipped because they don't evaluate on x86_64-linux: pkgsCross.arduino-avr.hello

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `pkgsCross.arduino-avr.hello' because:
error: attribute 'arduino-avr' in selection path 'pkgsCross.arduino-avr.hello' not found

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=510da5d5-f366-4c6f-aeb4-d8913af98654-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=4019d31d-0dd6-4106-a8df-eb4e3a21bf5e)

Attempted: pkgsCross.arduino-uno.hello

<details><summary>Partial log (click to expand)</summary><p>

```
 #include_next <limits.h>  /* recurse down to the real one */
               ^~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:1658: lib/c-strcasecmp.o] Error 1
make[2]: Leaving directory '/build/hello-2.10'
make[1]: *** [Makefile:1878: all-recursive] Error 1
make[1]: Leaving directory '/build/hello-2.10'
make: *** [Makefile:1413: all] Error 2
builder for '/nix/store/w5bk6jp2fwc8yhz9zzha4k8kv8mj54df-hello-2.10-avr.drv' failed with exit code 2
error: build of '/nix/store/w5bk6jp2fwc8yhz9zzha4k8kv8mj54df-hello-2.10-avr.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=510da5d5-f366-4c6f-aeb4-d8913af98654-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=0b75cf83-5b2e-40c9-9716-85c32addcf05)

Attempted: pkgsCross.arduino-uno.hello

<details><summary>Partial log (click to expand)</summary><p>

```
 #include_next <limits.h>  /* recurse down to the real one */
               ^~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:1658: lib/c-strcasecmp.o] Error 1
make[2]: Leaving directory '/private/tmp/nix-build-hello-2.10-avr.drv-0/hello-2.10'
make[1]: *** [Makefile:1878: all-recursive] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-hello-2.10-avr.drv-0/hello-2.10'
make: *** [Makefile:1413: all] Error 2
builder for '/nix/store/kvl688caam66mqxpq5whqbdwxmckg8qi-hello-2.10-avr.drv' failed with exit code 2
error: build of '/nix/store/kvl688caam66mqxpq5whqbdwxmckg8qi-hello-2.10-avr.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=510da5d5-f366-4c6f-aeb4-d8913af98654-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=e2ca4fe4-3067-4194-bd08-149d8b3389fb)

Attempted: pkgsCross.arduino-uno.hello

<details><summary>Partial log (click to expand)</summary><p>

```
 #include_next <limits.h>  /* recurse down to the real one */
               ^~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:1658: lib/c-strcasecmp.o] Error 1
make[2]: Leaving directory '/build/hello-2.10'
make[1]: *** [Makefile:1878: all-recursive] Error 1
make[1]: Leaving directory '/build/hello-2.10'
make: *** [Makefile:1413: all] Error 2
builder for '/nix/store/674083hq20pvkiz97c5a0yc5hnlq5hv7-hello-2.10-avr.drv' failed with exit code 2
error: build of '/nix/store/674083hq20pvkiz97c5a0yc5hnlq5hv7-hello-2.10-avr.drv' failed
```
</p></details>

Ericson2314 commented:
Oh I recall our GCC derivation doing something funny with limits.h

matthewbauer commented:
@ericson2314 It looks like it's coming from this line:

https://github.com/nixos/nixpkgs/blob/master/pkgs/development/compilers/gcc/builder.sh#L142

We've always assumed we had limits.h, but that's not always the case. Maybe remove that line from the builder? (later on of course, trying to avoid mass rebuilds here)

matthewbauer commented:
That is we overwrite the builtin LIMITS_H_TEST with that flag, causing it to always be true:

https://github.com/gcc-mirror/gcc/blob/master/gcc/Makefile.in#L535

But it looks like we can't avoid the mass rebuild unfortunately!

matthewbauer committed:
release-cross: add new targets

adds:

- arm-embedded
- avr
- aarch64-embedded
- powerpc-embedded

GrahamcOfBorg commented:
<!--REQUEST_ID=953479d9-14b3-4f7d-96b8-70011a296665-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=3c0307ad-9b23-4342-901e-66cbba33d04f)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on x86_64-darwin: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```

these paths will be fetched (37.20 MiB download, 155.10 MiB unpacked):
  /nix/store/10n6qiaz1piz0gsh05x5i062cscrwi6f-gcc-wrapper-7.3.0
  /nix/store/6d95j619zw7n6sbk5dqs36gwrsvif9x0-gcc-7.3.0
  /nix/store/alrfpp3mnyr5lycdh1xi36mhpclnkv0l-gcc-7.3.0-lib
copying path '/nix/store/alrfpp3mnyr5lycdh1xi36mhpclnkv0l-gcc-7.3.0-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/6d95j619zw7n6sbk5dqs36gwrsvif9x0-gcc-7.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/10n6qiaz1piz0gsh05x5i062cscrwi6f-gcc-wrapper-7.3.0' from 'https://cache.nixos.org'...
/nix/store/10n6qiaz1piz0gsh05x5i062cscrwi6f-gcc-wrapper-7.3.0
/nix/store/7l83571b34mmbr1zkxa596vw2wdqxxnh-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=953479d9-14b3-4f7d-96b8-70011a296665-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=d9118785-a8b9-41e6-a82c-20a4c7abfecc)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on aarch64-linux: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/klgxdd526njckk7cq8j1pxakb2p6wv70-gcc-wrapper-7.3.0
/nix/store/m6llsl31vkhkhxfy055m44444nz9s9ki-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=953479d9-14b3-4f7d-96b8-70011a296665-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=673def7f-33ea-49c0-922c-66d63d176ef0)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on x86_64-linux: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/10yq7kwlvbc6h658izmrlsspry1g9f3c-gcc-wrapper-7.3.0
/nix/store/glnk048cqva0ig8yrjfw140z2r2v18b4-stdenv-linux
```
</p></details>

matthewbauer commented:
@GrahamcOfBorg build pkgsCross.arduino-uno.hello

GrahamcOfBorg commented:
<!--REQUEST_ID=e62bd24c-5519-4b73-8fb2-fb62154ccae8-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=c9a00297-603a-4261-bff8-8c4cdfb47320)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on x86_64-linux: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/10yq7kwlvbc6h658izmrlsspry1g9f3c-gcc-wrapper-7.3.0
/nix/store/glnk048cqva0ig8yrjfw140z2r2v18b4-stdenv-linux
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=e62bd24c-5519-4b73-8fb2-fb62154ccae8-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=d928476a-5f63-4eb8-9883-a839bd95b34c)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on aarch64-linux: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/klgxdd526njckk7cq8j1pxakb2p6wv70-gcc-wrapper-7.3.0
/nix/store/m6llsl31vkhkhxfy055m44444nz9s9ki-stdenv-linux
```
</p></details>

matthewbauer commented:
Pinging a few people who have used avrgcc to see what can be done to make this transition easier:

Specifically how can we get multiple compilers to work with cc-wrapper? Without knowing ICU and MCU ahead of time we are kind of stuck...

- @mguentner
- @acowley 
- @gesturbjarkason

GrahamcOfBorg commented:
<!--REQUEST_ID=2537f1be-6635-40b8-827b-585cc81833fc-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=1bc7c71e-684a-4a85-8d10-3596bab167ba)

Attempted: pkgsCross.arduino-uno.hello

<details><summary>Partial log (click to expand)</summary><p>

```
                   close (fd);
                   ^~~~~
                   fclose
make[2]: *** [Makefile:1658: lib/localcharset.o] Error 1
make[2]: Leaving directory '/build/hello-2.10'
make[1]: *** [Makefile:1878: all-recursive] Error 1
make[1]: Leaving directory '/build/hello-2.10'
make: *** [Makefile:1413: all] Error 2
builder for '/nix/store/sk9rxds3a6y7pfkkjynh716ibcsbw1rf-hello-2.10-avr.drv' failed with exit code 2
error: build of '/nix/store/sk9rxds3a6y7pfkkjynh716ibcsbw1rf-hello-2.10-avr.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=2537f1be-6635-40b8-827b-585cc81833fc-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=75782cbe-3e36-4157-900b-fc7be7df1c40)

Attempted: pkgsCross.arduino-uno.hello

<details><summary>Partial log (click to expand)</summary><p>

```
                   close (fd);
                   ^~~~~
                   fclose
make[2]: *** [Makefile:1658: lib/localcharset.o] Error 1
make[2]: Leaving directory '/build/hello-2.10'
make[1]: *** [Makefile:1878: all-recursive] Error 1
make[1]: Leaving directory '/build/hello-2.10'
make: *** [Makefile:1413: all] Error 2
builder for '/nix/store/i61sx5li0lqaia39mb73mwm7xq97wwi3-hello-2.10-avr.drv' failed with exit code 2
error: build of '/nix/store/i61sx5li0lqaia39mb73mwm7xq97wwi3-hello-2.10-avr.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=e62bd24c-5519-4b73-8fb2-fb62154ccae8-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=48dbb849-06ea-4880-88aa-982fcc2e63e1)

Attempted: gcc, stdenv

The following builds were skipped because they don't evaluate on x86_64-darwin: avrdudess

<details><summary>Partial log (click to expand)</summary><p>

```
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/10n6qiaz1piz0gsh05x5i062cscrwi6f-gcc-wrapper-7.3.0
/nix/store/7l83571b34mmbr1zkxa596vw2wdqxxnh-stdenv-darwin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=2537f1be-6635-40b8-827b-585cc81833fc-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=b487e30b-4088-4fa2-b278-15c8e1a24b76)

Attempted: pkgsCross.arduino-uno.hello

<details><summary>Partial log (click to expand)</summary><p>

```
                   close (fd);
                   ^~~~~
                   fclose
make[2]: *** [Makefile:1658: lib/localcharset.o] Error 1
make[2]: Leaving directory '/private/tmp/nix-build-hello-2.10-avr.drv-0/hello-2.10'
make[1]: *** [Makefile:1878: all-recursive] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-hello-2.10-avr.drv-0/hello-2.10'
make: *** [Makefile:1413: all] Error 2
builder for '/nix/store/djcfv74bd24walc9z8kdnhf2qj5kz3f1-hello-2.10-avr.drv' failed with exit code 2
error: build of '/nix/store/djcfv74bd24walc9z8kdnhf2qj5kz3f1-hello-2.10-avr.drv' failed
```
</p></details>

matthewbauer committed:
dfu-programmer: supports macos

matthewbauer committed:
newlib: init at 3.0.0

Fixes #47533

newlib seems to think it’s a compiler when it’s not. Anyway had to
make host refer to build platform for things to work.

matthewbauer committed:
systems/parse.nix: support weird system configs

There are some weird ones out there that don’t follow any pattern:

- arm-none-eabi
- powerpc-none-eabi
- aarch64-none-elf

matthewbauer committed:
gcc-arm-embedded: remove

replace with:

- pkgsCross.arm-embedded.buildPackages.gcc
- pkgsCross.arm-embedded.buildPackages.binutils

matthewbauer committed:
mentorToolchains: remove

mentorToolchains.armLinuxGnuEabi is replaced with:

- pkgsCross.pogoplug4.buildPackages.gcc
- pkgsCross.pogoplug4.buildPackages.binutils

and mentorToolchains.armEabi is replaced with:

- pkgsCross.arm-embedded.buildPackages.gcc
- pkgsCross.arm-embedded.buildPackages.binutils

Fixes #39087

matthewbauer committed:
qmk_firmware: init at 0.6.144

A convenient way to test embedded projects

matthewbauer committed:
gnuk{-git,-unstable}: remove

These two are unmaintained, no reason to keep them when they are
out-of-date

GrahamcOfBorg commented:
<!--REQUEST_ID=59a564e2-da79-4296-939e-4c64c4fc11b3-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=013d1364-336d-4d29-a781-3c60a2093240)

Attempted: dfu-programmer, gcc, newlib, stdenv

The following builds were skipped because they don't evaluate on x86_64-darwin: avrdudess, opentx, qmk_firmware

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: Resource temporarily unavailable
builder for '/nix/store/0782dybi3wk39b41j447mpc2cdfxh9ni-curl-7.61.1.drv' failed with exit code 254
cannot build derivation '/nix/store/s25qjww0mz3rg1i6hc103v1y1w1a4vgw-newlib-3.0.0.tar.gz.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/fc5cz85aswfp4mqf0z6pyidbbalhwg1r-newlib-3.0.0.drv': 1 dependencies couldn't be built
builder for '/nix/store/pzi01l4dxgqc06z3aq68vyhnv83qvxfw-dfu-programmer-0.7.2.drv' failed with exit code 254
error: build of '/nix/store/fc5cz85aswfp4mqf0z6pyidbbalhwg1r-newlib-3.0.0.drv', '/nix/store/pzi01l4dxgqc06z3aq68vyhnv83qvxfw-dfu-programmer-0.7.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=59a564e2-da79-4296-939e-4c64c4fc11b3-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=f510f3c8-79ea-482f-bdfb-e6631f842779)

Attempted: dfu-programmer, gcc, newlib, qmk_firmware, stdenv

The following builds were skipped because they don't evaluate on aarch64-linux: avrdudess, opentx

<details><summary>Partial log (click to expand)</summary><p>

```
make[1]: *** [tmk_core/rules.mk:286: .build/z150_blackheart_default.elf] Error 1
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/g6kdngcpb0q0cja3h82413ms3ypi2d39-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/g6kdngcpb0q0cja3h82413ms3ypi2d39-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=59a564e2-da79-4296-939e-4c64c4fc11b3-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=fba05647-53ce-4d38-bb03-06f9b0632146)

Attempted: avrdudess, dfu-programmer, gcc, newlib, opentx, qmk_firmware, stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
Making vitamins_included/rev1 with keymap default                                                      [OK]
Making viterbi/rev1 with keymap default                                                                [OK]
Making viterbi/rev1 with keymap default                                                                [OK]
Making wavelet with keymap default                                                                     [OK]
Making whitefox with keymap default                                                                    [WARNINGS]
(Firmware size check does not yet support cortex-m4 microprocessors; skipping.)
Making wilba_tech/wt60_a with keymap default                                                           [OK]
Making wilba_tech/wt65_a with keymap default                                                           [OK]
building of '/nix/store/16nq0h56krjw96balhx2qh3k02yimdws-qmk_firmware-0.6.144.drv' timed out after 3600 seconds
error: build of '/nix/store/16nq0h56krjw96balhx2qh3k02yimdws-qmk_firmware-0.6.144.drv', '/nix/store/2rj8qpc4diy91cx0w2zn1dbwx8la3kqv-opentx-2.2.1.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=ebef4f6c-698e-4dfe-8830-1a78f74f06a7-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=d3c74c6b-51db-4d37-a9e9-9d64f4d77ee2)

Attempted: dfu-programmer, gcc, newlib, stdenv

The following builds were skipped because they don't evaluate on x86_64-darwin: avrdudess, opentx, qmk_firmware

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: retry: Resource temporarily unavailable
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin/setup: fork: Resource temporarily unavailable
builder for '/nix/store/pzi01l4dxgqc06z3aq68vyhnv83qvxfw-dfu-programmer-0.7.2.drv' failed with exit code 254
builder for '/nix/store/0782dybi3wk39b41j447mpc2cdfxh9ni-curl-7.61.1.drv' failed with exit code 254
cannot build derivation '/nix/store/s25qjww0mz3rg1i6hc103v1y1w1a4vgw-newlib-3.0.0.tar.gz.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/fc5cz85aswfp4mqf0z6pyidbbalhwg1r-newlib-3.0.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/fc5cz85aswfp4mqf0z6pyidbbalhwg1r-newlib-3.0.0.drv', '/nix/store/pzi01l4dxgqc06z3aq68vyhnv83qvxfw-dfu-programmer-0.7.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=ebef4f6c-698e-4dfe-8830-1a78f74f06a7-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=e0a01805-cae9-4679-8fb2-2e0df6ed6a72)

Attempted: avrdudess, dfu-programmer, gcc, newlib, opentx, qmk_firmware, stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/mqmr0syag2p7ch8lddnnph5sjvhml8wc-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/2rj8qpc4diy91cx0w2zn1dbwx8la3kqv-opentx-2.2.1.drv', '/nix/store/mqmr0syag2p7ch8lddnnph5sjvhml8wc-qmk_firmware-0.6.144.drv' failed
```
</p></details>

matthewbauer committed:
teensy-loader-cli: build on macos

GrahamcOfBorg commented:
<!--REQUEST_ID=b0c7b5b5-f3c3-422e-80f5-f107d860e1b3-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=98731c12-76b3-4327-a976-ca3638ecf49e)

Attempted: dfu-programmer, gcc, newlib, stdenv

The following builds were skipped because they don't evaluate on x86_64-darwin: avrdudess, opentx, qmk_firmware

<details><summary>Partial log (click to expand)</summary><p>

```
make[2]: Leaving directory '/private/tmp/nix-build-newlib-3.0.0.drv-0/newlib-3.0.0/etc'
make[1]: Nothing to be done for 'install-target'.
make[1]: Leaving directory '/private/tmp/nix-build-newlib-3.0.0.drv-0/newlib-3.0.0'
post-installation fixup
strip is /nix/store/g5r4apl0za012ffs6ladinwa5w0m1l3k-cctools-binutils-darwin/bin/strip
patching script interpreter paths in /nix/store/ggz078k77gdn4j95ld64lzwrqxbh41dr-newlib-3.0.0
/nix/store/qx9jfm08jrc1ig02icfgknph3i9rwny1-dfu-programmer-0.7.2
/nix/store/xs8x90rif8vp3ymhdyqpc66cx0l7s2c9-gcc-wrapper-7.3.0
/nix/store/ggz078k77gdn4j95ld64lzwrqxbh41dr-newlib-3.0.0
/nix/store/8qai18dv0nqyw6s61g9p91n8684rnlg8-stdenv-darwin
```
</p></details>

matthewbauer commented:
Just squashed commits and cleaned things up. Interested in getting any feedback on this!

GrahamcOfBorg commented:
<!--REQUEST_ID=06b9949c-b9d2-4748-aa8d-9b70814b273b-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=50b519d8-1877-46e7-89b9-09e389fb047f)

Attempted: qmk_firmware

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/kb0av5cx9byxw230k74y6hljk2asqz2f-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/kb0av5cx9byxw230k74y6hljk2asqz2f-qmk_firmware-0.6.144.drv' failed
```
</p></details>

Ericson2314 commented on pkgs/applications/audio/axoloti/default.nix:
```diff
@@ -20,7 +21,7 @@ stdenv.mkDerivation rec {
   };
 
-  buildInputs = [ makeWrapper unzip gcc-arm-embedded dfu-util-axoloti jdk ant libfaketime ];
+  buildInputs = [ makeWrapper unzip gcc-arm-embedded binutils-arm-embedded dfu-util-axoloti jdk ant libfaketime ];
```
Some of these should be nativeBuildInputs but whatever, one problem at a time :)

GrahamcOfBorg commented:
<!--REQUEST_ID=06b9949c-b9d2-4748-aa8d-9b70814b273b-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=256d4dba-8547-4955-b90f-6ea2324f5cc7)

Attempted: qmk_firmware

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/ncfy8l7ckdg57ybj76y84asl54ql5qz8-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/ncfy8l7ckdg57ybj76y84asl54ql5qz8-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=b0c7b5b5-f3c3-422e-80f5-f107d860e1b3-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=de01e4c4-5a5b-41c6-abae-d11596e2e201)

Attempted: avrdudess, dfu-programmer, gcc, newlib, opentx, qmk_firmware, stdenv

<details><summary>Partial log (click to expand)</summary><p>

```
[ 37%] Linking CXX executable ../../firmware.elf
/nix/store/z6pvjan22wqxcg787kyw89jl64dhh2ks-arm-none-eabi-binutils-2.30/bin/arm-none-eabi-ld: cannot find -lstdc++_nano
/nix/store/z6pvjan22wqxcg787kyw89jl64dhh2ks-arm-none-eabi-binutils-2.30/bin/arm-none-eabi-ld: cannot find -lg_nano
/nix/store/z6pvjan22wqxcg787kyw89jl64dhh2ks-arm-none-eabi-binutils-2.30/bin/arm-none-eabi-ld: cannot find -lc_nano
collect2: error: ld returned 1 exit status
make[2]: *** [radio/src/CMakeFiles/firmware.dir/build.make:2581: firmware.elf] Error 1
make[1]: *** [CMakeFiles/Makefile2:148: radio/src/CMakeFiles/firmware.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
builder for '/nix/store/2rj8qpc4diy91cx0w2zn1dbwx8la3kqv-opentx-2.2.1.drv' failed with exit code 2
error: build of '/nix/store/2rj8qpc4diy91cx0w2zn1dbwx8la3kqv-opentx-2.2.1.drv', '/nix/store/mqmr0syag2p7ch8lddnnph5sjvhml8wc-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=06b9949c-b9d2-4748-aa8d-9b70814b273b-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=76ab2ac6-2bb1-4fee-98f4-557189ed3914)

Attempted: qmk_firmware

<details><summary>Partial log (click to expand)</summary><p>

```
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
make: *** [Makefile:544: all:default] Error 1
Make finished with errors

builder for '/nix/store/d5wmab3pgdzpq6s40kxmmbcxqrddm7g1-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/d5wmab3pgdzpq6s40kxmmbcxqrddm7g1-qmk_firmware-0.6.144.drv' failed
```
</p></details>

danielfullmer commented:
Very cool. Thanks for working on this. I can at least confirm this works with my own custom keyboard firmware based on `qmk_firmware`. If interested, see: https://github.com/danielfullmer/nixos-config/commit/27447be5cc69c30b44e34ec379130183a3d16a5d

matthewbauer committed:
axoloti: use nativeBuildInputs correctly

GrahamcOfBorg commented:
<!--REQUEST_ID=989f1b4d-14c7-4465-83bf-ce22b7af9641-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=df95eff8-f8a2-45ee-9fa8-ddf6c5cc353c)

Attempted: axoloti, dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

<details><summary>Partial log (click to expand)</summary><p>

```
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
make: *** [Makefile:544: all:default] Error 1
Make finished with errors

builder for '/nix/store/wy6nxg0j9lhq64b8ad6nsaz8a1llml63-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/nwmnccq9dpdgy6qdw6l5j85124zl131p-axoloti-1.0.12-2.drv', '/nix/store/wy6nxg0j9lhq64b8ad6nsaz8a1llml63-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=989f1b4d-14c7-4465-83bf-ce22b7af9641-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=77aca675-dfe9-4ab0-8854-48d4ce7487d4)

Attempted: dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

The following builds were skipped because they don't evaluate on aarch64-linux: axoloti

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/9mmwbcrfr0yk02097ymc60ag4smk4h22-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/9mmwbcrfr0yk02097ymc60ag4smk4h22-qmk_firmware-0.6.144.drv' failed
```
</p></details>

matthewbauer committed:
systems/examples: add i686 & x86-64 embedded

Fixes #28160

matthewbauer commented:
Merging this now. Apologies if this breaks anything. I will be available to fix any potential breakages this weekend. Please report any issues as the appear!

GrahamcOfBorg commented:
<!--REQUEST_ID=30ac1bcd-f4bf-4aea-999b-d967c63573e4-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=50b0c890-6cad-4d5d-b3a5-76f1227c3b20)

Attempted: axoloti, dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/n0jxmmxhdxzdbr8sxilmkrki5pb9yjbn-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/bp1lnx8mwhbn1gs6dzp2vl3y5qw8psbb-axoloti-1.0.12-2.drv', '/nix/store/n0jxmmxhdxzdbr8sxilmkrki5pb9yjbn-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=30ac1bcd-f4bf-4aea-999b-d967c63573e4-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=51601804-6b44-4b7a-ad61-3873050ff15c)

Attempted: dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

The following builds were skipped because they don't evaluate on aarch64-linux: axoloti

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/9mmwbcrfr0yk02097ymc60ag4smk4h22-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/9mmwbcrfr0yk02097ymc60ag4smk4h22-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=551b1741-e3b8-44c1-af70-6d0e4361c414-->
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=d2e3e937-eed7-41f6-ab89-dc5aba9eee55)

Attempted: axoloti, dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

<details><summary>Partial log (click to expand)</summary><p>

```
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
make: *** [Makefile:544: all:default] Error 1
Make finished with errors

builder for '/nix/store/wy6nxg0j9lhq64b8ad6nsaz8a1llml63-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/nwmnccq9dpdgy6qdw6l5j85124zl131p-axoloti-1.0.12-2.drv', '/nix/store/wy6nxg0j9lhq64b8ad6nsaz8a1llml63-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=989f1b4d-14c7-4465-83bf-ce22b7af9641-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=e1446289-048e-412b-8b25-551425804c40)

Attempted: axoloti, dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/1v2gny7fnmzn06ixv61hlwazdx4h52sb-arm-none-eabi-stage-final-gcc-debug-wrapper-7.3.0/bin/arm-none-eabi-g++ -c -mcpu=cortex-m4 -O2 -ggdb -fomit-frame-pointer -falign-functions=16 -fno-math-errno -Wno-unused-parameter -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsingle-precision-constant -fno-rtti -Wall -Wextra -Wa,-alms=build/lst/svf.lst  -DCORTEX_USE_FPU=TRUE -DARM_MATH_CM4 -D__FPU_PRESENT -DTHUMB_PRESENT -mno-thumb-interwork -DTHUMB_NO_INTERWORKING -MD -MP -MF .dep/svf.o.d -mthumb -DTHUMB -I. -I/build/source/platform_linux/../CMSIS//Include -I/build/source/platform_linux/../CMSIS//DSP_Lib/Include -I/build/source/platform_linux/../chibios/os/ports/common/ARMCMx/CMSIS/include -I/build/source/platform_linux/../chibios/os/ports/common/ARMCMx -I/build/source/platform_linux/../chibios/os/ports/GCC/ARMCMx -I/build/source/platform_linux/../chibios/os/ports/GCC/ARMCMx/STM32F4xx -I/build/source/platform_linux/../chibios/os/kernel/include -I/build/source/platform_linux/../chibios/os/hal/include -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32F4xx -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/GPIOv2 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/I2Cv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/OTGv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/RTCv2 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/SPIv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/TIMv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/USARTv1 -I. -I/build/source/platform_linux/../chibios/ext/fatfs/src -I/build/source/platform_linux/../chibios/os/various -I./STM32_USB_Host_Library/Core/Inc -I./STM32_USB_Host_Library/Class/HID/Inc -I./STM32F4xx_HAL_Driver/Inc -I./mutable_instruments -I./mutable_instruments/warps -I./mutable_instruments/braids -I./mutable_instruments/rings -I./mutable_instruments/elements -I./mutable_instruments/stmlib ./mutable_instruments/streams/svf.cpp -o build/obj/svf.o

/nix/store/1v2gny7fnmzn06ixv61hlwazdx4h52sb-arm-none-eabi-stage-final-gcc-debug-wrapper-7.3.0/bin/arm-none-eabi-g++ -c -mcpu=cortex-m4 -O2 -ggdb -fomit-frame-pointer -falign-functions=16 -fno-math-errno -Wno-unused-parameter -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsingle-precision-constant -fno-rtti -Wall -Wextra -Wa,-alms=build/lst/vactrol.lst  -DCORTEX_USE_FPU=TRUE -DARM_MATH_CM4 -D__FPU_PRESENT -DTHUMB_PRESENT -mno-thumb-interwork -DTHUMB_NO_INTERWORKING -MD -MP -MF .dep/vactrol.o.d -mthumb -DTHUMB -I. -I/build/source/platform_linux/../CMSIS//Include -I/build/source/platform_linux/../CMSIS//DSP_Lib/Include -I/build/source/platform_linux/../chibios/os/ports/common/ARMCMx/CMSIS/include -I/build/source/platform_linux/../chibios/os/ports/common/ARMCMx -I/build/source/platform_linux/../chibios/os/ports/GCC/ARMCMx -I/build/source/platform_linux/../chibios/os/ports/GCC/ARMCMx/STM32F4xx -I/build/source/platform_linux/../chibios/os/kernel/include -I/build/source/platform_linux/../chibios/os/hal/include -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32F4xx -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/GPIOv2 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/I2Cv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/OTGv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/RTCv2 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/SPIv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/TIMv1 -I/build/source/platform_linux/../chibios/os/hal/platforms/STM32/USARTv1 -I. -I/build/source/platform_linux/../chibios/ext/fatfs/src -I/build/source/platform_linux/../chibios/os/various -I./STM32_USB_Host_Library/Core/Inc -I./STM32_USB_Host_Library/Class/HID/Inc -I./STM32F4xx_HAL_Driver/Inc -I./mutable_instruments -I./mutable_instruments/warps -I./mutable_instruments/braids -I./mutable_instruments/rings -I./mutable_instruments/elements -I./mutable_instruments/stmlib ./mutable_instruments/streams/vactrol.cpp -o build/obj/vactrol.o

/nix/store/1v2gny7fnmzn06ixv61hlwazdx4h52sb-arm-none-eabi-stage-final-gcc-debug-wrapper-7.3.0/bin/arm-none-eabi-gcc   build/obj/arm_bitreversal2.o  build/obj/crt0.o build/obj/vectors.o build/obj/chcore.o build/obj/chcore_v7m.o build/obj/nvic.o build/obj/chsys.o build/obj/chdebug.o build/obj/chlists.o build/obj/chvt.o build/obj/chschd.o build/obj/chthreads.o build/obj/chdynamic.o build/obj/chregistry.o build/obj/chsem.o build/obj/chmtx.o build/obj/chcond.o build/obj/chevents.o build/obj/chmsg.o build/obj/chmboxes.o build/obj/chqueues.o build/obj/chmemcore.o build/obj/chheap.o build/obj/chmempools.o build/obj/hal.o build/obj/adc.o build/obj/can.o build/obj/ext.o build/obj/gpt.o build/obj/i2c.o build/obj/icu.o build/obj/mac.o build/obj/mmc_spi.o build/obj/mmcsd.o build/obj/pal.o build/obj/pwm.o build/obj/rtc.o build/obj/sdc.o build/obj/serial.o build/obj/serial_usb.o build/obj/spi.o build/obj/tm.o build/obj/uart.o build/obj/usb.o build/obj/stm32_dma.o build/obj/hal_lld.o build/obj/adc_lld.o build/obj/ext_lld_isr.o build/obj/can_lld.o build/obj/ext_lld.o build/obj/mac_lld.o build/obj/sdc_lld.o build/obj/pal_lld.o build/obj/i2c_lld.o build/obj/usb_lld.o build/obj/rtc_lld.o build/obj/spi_lld.o build/obj/gpt_lld.o build/obj/icu_lld.o build/obj/pwm_lld.o build/obj/serial_lld.o build/obj/uart_lld.o build/obj/board.o build/obj/fatfs_diskio.o build/obj/fatfs_syscall.o build/obj/ff.o build/obj/ccsbcs.o build/obj/chprintf.o build/obj/memstreams.o build/obj/axoloti_control.o build/obj/codec.o build/obj/axoloti.o build/obj/axoloti_filters.o build/obj/axoloti_math.o build/obj/axoloti_oscs.o build/obj/axoloti_board.o build/obj/midi.o build/obj/ui.o build/obj/glcdfont.o build/obj/usbcfg.o build/obj/sdcard.o build/obj/pconnection.o build/obj/parameters.o build/obj/main.o build/obj/patch.o build/obj/flash.o build/obj/virtual_control.o build/obj/arm_bitreversal.o build/obj/arm_cfft_f32.o build/obj/arm_cfft_q15.o build/obj/arm_cfft_q31.o build/obj/arm_cfft_radix2_f32.o build/obj/arm_cfft_radix2_init_f32.o build/obj/arm_cfft_radix2_init_q15.o build/obj/arm_cfft_radix2_init_q31.o build/obj/arm_cfft_radix2_q15.o build/obj/arm_cfft_radix2_q31.o build/obj/arm_cfft_radix4_f32.o build/obj/arm_cfft_radix4_init_f32.o build/obj/arm_cfft_radix4_init_q15.o build/obj/arm_cfft_radix4_init_q31.o build/obj/arm_cfft_radix4_q15.o build/obj/arm_cfft_radix4_q31.o build/obj/arm_cfft_radix8_f32.o build/obj/arm_rfft_f32.o build/obj/arm_rfft_fast_f32.o build/obj/arm_rfft_fast_init_f32.o build/obj/arm_rfft_init_f32.o build/obj/arm_rfft_init_q15.o build/obj/arm_rfft_init_q31.o build/obj/arm_rfft_q15.o build/obj/arm_rfft_q31.o build/obj/arm_mult_q31.o build/obj/arm_mult_q15.o build/obj/arm_mult_f32.o build/obj/arm_shift_q31.o build/obj/arm_shift_q15.o build/obj/arm_scale_f32.o build/obj/arm_cmplx_mult_cmplx_q31.o build/obj/arm_cmplx_mult_cmplx_f32.o build/obj/arm_cmplx_mag_q31.o build/obj/arm_cmplx_mult_cmplx_q15.o build/obj/arm_cmplx_mag_q15.o build/obj/arm_cos_q31.o build/obj/arm_sin_q31.o build/obj/arm_sqrt_q31.o build/obj/arm_sqrt_q15.o build/obj/arm_fir_init_q31.o build/obj/arm_fir_fast_q31.o build/obj/arm_biquad_cascade_df1_32x64_init_q31.o build/obj/arm_biquad_cascade_df1_32x64_q31.o build/obj/arm_biquad_cascade_df1_fast_q31.o build/obj/arm_biquad_cascade_df1_init_q31.o build/obj/arm_biquad_cascade_df1_q31.o build/obj/arm_fir_interpolate_init_q31.o build/obj/arm_fir_interpolate_q31.o build/obj/arm_fir_decimate_init_q31.o build/obj/arm_fir_decimate_q31.o build/obj/arm_fir_decimate_fast_q31.o build/obj/arm_fir_lattice_init_q31.o build/obj/arm_fir_lattice_q31.o build/obj/arm_iir_lattice_init_q31.o build/obj/arm_iir_lattice_q31.o build/obj/arm_common_tables.o build/obj/arm_const_structs.o build/obj/usbh_core.o build/obj/usbh_ctlreq.o build/obj/usbh_ioreq.o build/obj/usbh_pipes.o build/obj/usbh_hid.o build/obj/usbh_hid_keybd.o build/obj/usbh_hid_mouse.o build/obj/usbh_hid_parser.o build/obj/stm32f4xx_ll_usb.o build/obj/stm32f4xx_hal_gpio.o build/obj/stm32f4xx_hal_cortex.o build/obj/stm32f4xx_hal_hcd.o build/obj/usbh_midi_core.o build/obj/usbh_vendor.o build/obj/usbh_conf.o build/obj/serial_midi.o build/obj/bulk_usb.o build/obj/midi_usb.o build/obj/crc32.o build/obj/exceptions.o build/obj/watchdog.o build/obj/sysmon.o build/obj/axoloti_memory.o  build/obj/units.o build/obj/atan.o build/obj/random.o build/obj/mutable_resources.o build/obj/braids_resources.o build/obj/analog_oscillator.o build/obj/digital_oscillator.o build/obj/macro_oscillator.o build/obj/axoloti_mi.o build/obj/multistage_envelope.o build/obj/ominous_voice.o build/obj/resonator.o build/obj/string.o build/obj/tube.o build/obj/exciter.o build/obj/elements_resources.o build/obj/part.o build/obj/voice.o build/obj/clouds_resources.o build/obj/correlator.o build/obj/mu_law.o build/obj/granular_processor.o build/obj/frame_transformation.o build/obj/phase_vocoder.o build/obj/stft.o build/obj/rings_resources.o build/obj/streams_resources.o build/obj/compressor.o build/obj/envelope.o build/obj/follower.o build/obj/lorenz_generator.o build/obj/processor.o build/obj/svf.o build/obj/vactrol.o -mcpu=cortex-m4 -O2 -ggdb -fomit-frame-pointer -falign-functions=16 -fno-math-errno -Wno-unused-parameter -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsingle-precision-constant -nostartfiles -L. -Wl,-Map=build/axoloti.map,--cref,--no-warn-mismatch,--library-path=/build/source/platform_linux/../chibios/os/ports/GCC/ARMCMx,--script=STM32F407xG.ld -mno-thumb-interwork -mthumb  -lm  -o build/axoloti.elf
/nix/store/1v2gny7fnmzn06ixv61hlwazdx4h52sb-arm-none-eabi-stage-final-gcc-debug-wrapper-7.3.0/bin/arm-none-eabi-objcopy -O ihex build/axoloti.elf build/axoloti.hex
make: /nix/store/1v2gny7fnmzn06ixv61hlwazdx4h52sb-arm-none-eabi-stage-final-gcc-debug-wrapper-7.3.0/bin/arm-none-eabi-objcopy: Command not found
make: *** [/build/source/platform_linux/../chibios/os/ports/GCC/ARMCMx/rules.mk:216: build/axoloti.hex] Error 127
builder for '/nix/store/bp1lnx8mwhbn1gs6dzp2vl3y5qw8psbb-axoloti-1.0.12-2.drv' failed with exit code 1
error: build of '/nix/store/bp1lnx8mwhbn1gs6dzp2vl3y5qw8psbb-axoloti-1.0.12-2.drv', '/nix/store/n0jxmmxhdxzdbr8sxilmkrki5pb9yjbn-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=30ac1bcd-f4bf-4aea-999b-d967c63573e4-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=9ce6c62c-9689-43e6-ab73-f15df53a988d)

Attempted: axoloti, dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

<details><summary>Partial log (click to expand)</summary><p>

```
 ^~~~~~~~
/nix/store/lxi68avn1z2ihd4xxma4wl7y5ws19sg3-libc++-5.0.2/include/c++/v1/cmath:589:1: error: template with C linkage
 template <class _A1>
 ^~~~~~~~
/nix/store/lxi68avn1z2ihd4xxma4wl7y5ws19sg3-libc++-5.0.2/include/c++/v1/cmath:601:1: error: template with C linkage
 template <class _A1>
 ^~~~~~~~
make: *** [/private/tmp/nix-build-axoloti-1.0.12-2.drv-0/source/platform_linux/../chibios/os/ports/GCC/ARMCMx/rules.mk:163: build/obj/units.o] Error 1
builder for '/nix/store/nwmnccq9dpdgy6qdw6l5j85124zl131p-axoloti-1.0.12-2.drv' failed with exit code 1
error: build of '/nix/store/nwmnccq9dpdgy6qdw6l5j85124zl131p-axoloti-1.0.12-2.drv', '/nix/store/wy6nxg0j9lhq64b8ad6nsaz8a1llml63-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=551b1741-e3b8-44c1-af70-6d0e4361c414-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=f70c5365-0888-495c-8cd3-b615187647a5)

Attempted: dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

The following builds were skipped because they don't evaluate on aarch64-linux: axoloti

<details><summary>Partial log (click to expand)</summary><p>

```
Making z150_blackheart with keymap default                                                             [OK]
Making zeal60 with keymap default                                                                      [OK]
Making zeal65 with keymap default                                                                      [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zen/rev1 with keymap default                                                                    [OK]
Making zlant with keymap default                                                                       [OK]
Make finished with errors
make: *** [Makefile:544: all:default] Error 1
builder for '/nix/store/9mmwbcrfr0yk02097ymc60ag4smk4h22-qmk_firmware-0.6.144.drv' failed with exit code 2
error: build of '/nix/store/9mmwbcrfr0yk02097ymc60ag4smk4h22-qmk_firmware-0.6.144.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=551b1741-e3b8-44c1-af70-6d0e4361c414-->
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48286&attempt_id=ebedf6db-533e-4951-bf21-d6d13efac025)

Attempted: axoloti, dfu-programmer, gcc, newlib, qmk_firmware, teensy-loader-cli

<details><summary>Partial log (click to expand)</summary><p>

```
/build/build/./gcc/xgcc -B/build/build/./gcc/   -msp8 -O2  -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE  -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -DDF=SF -Dinhibit_libc -mcall-prologues -Os -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector -Dinhibit_libc  -DDF=SF -Dinhibit_libc -mcall-prologues -Os -I. -I. -I../../.././gcc -I../../../../gcc-7.3.0/libgcc -I../../../../gcc-7.3.0/libgcc/. -I../../../../gcc-7.3.0/libgcc/../gcc -I../../../../gcc-7.3.0/libgcc/../include  -DHAVE_CC_TLS -DUSE_EMUTLS -o _satfractTIHQ.o -MT _satfractTIHQ.o -MD -MP -MF _satfractTIHQ.dep -DL_satfract -DFROM_TI -DTO_HQ -c ../../../../gcc-7.3.0/libgcc/fixed-bit.c -fvisibility=hidden -DHIDE_EXPORTS
building of '/nix/store/2c8dkbrx2zdi1gk1cb0wj3jh0jdpz6ms-avr-stage-static-gcc-debug-7.3.0.drv' timed out after 3600 seconds
cannot build derivation '/nix/store/wn03qv0zi813pcx0ppdzrfxlflcql3ff-avr-stage-static-gcc-debug-wrapper-7.3.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/qkpr63xyy1gs8hnw0s4znql5sjkcbwcy-stdenv-linux.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/jdd9bl0bg18nd0mvk0gn6mf59m4ch0a3-avr-libc-2.0.0-avr.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/1ca1397x6v5icikbis3ls9fihh9klrg0-avr-binutils-wrapper-2.30.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/54m1674ybfgvjgi7z4b6cdpm93lh3c3n-avr-stage-final-gcc-debug-7.3.0.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/fvhbpf0z7igcls0ca16n19pk0grhl8ad-avr-stage-final-gcc-debug-wrapper-7.3.0.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/n0jxmmxhdxzdbr8sxilmkrki5pb9yjbn-qmk_firmware-0.6.144.drv': 2 dependencies couldn't be built
error: build of '/nix/store/bp1lnx8mwhbn1gs6dzp2vl3y5qw8psbb-axoloti-1.0.12-2.drv', '/nix/store/n0jxmmxhdxzdbr8sxilmkrki5pb9yjbn-qmk_firmware-0.6.144.drv' failed
```
</p></details>

chloekek commented:
It was difficult for me to figure out how to find avr-gcc after this change, since command-not-found no longer reports which package provides it. Is there anything we can do about that?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-build-patched-avr-gcc-with-patched-avr-libc/21468/2
